### PR TITLE
Replace highlights and show (locally) stored annotations

### PR DESCRIPTION
**Pull request by andrenarchy**
This PR contains the following changes:

* remove rangy-highlight and rangy-classapplier
* new directives:
  * `onTextSelect`: capture text selections on an element
  * `highlight-container`, `highlight-root`, `highlight-target`: highlight selections
    * plays nicely with `onTextSelect`
    * can be applied to any DOM element (not restricted to PDFs)
    * does not modify the DOM subtree where the selections are made
  * `onOutsideClick`, `onOutsideMousedown`, `onOutsideMouseup`: capture mouse click/down/up *outside* of an element
* `pdf` directive: new `onLoaded` attribute

### CODE REVIEW

File: src/js/directives/highlights.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+  // returns all leaf text nodes that are descendants of node or are node
+  var getTextNodes = function (node) {
+    if (!node) return [];
+    if (node.nodeType === Node.TEXT_NODE) return [node];
+    
+    // process childs
+    var nodes = [];
+    $(node).contents().each(function (index, el) {
+      nodes = nodes.concat(getTextNodes(el));
```
Comment (nschloe) [inline src/js/directives/highlights.js]: `.push()`? [1] suggests that `concat` is faster, but then again we're doing this in-place 
[1] http://jsperf.com/javascript-array-concat-vs-push
[2] http://stackoverflow.com/questions/351409/appending-to-array

File: src/js/directives/highlights.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      template: '<div ng-transclude></div>',
+      controller: ['$scope', function ($scope) {
+      }]
+    };
+  });
+  
+  app.directive('highlightRoot', function () {
+    return {
+      restrict: 'A',
+      require: '^^highlightContainer',
```
Comment (nschloe) [inline src/js/directives/highlights.js]: What the?

File: src/js/directives/highlights.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      link: function (scope, element, attrs, containerCtrl) {
+        containerCtrl.getRangesRects = function (serializedRanges) {
+          if (!serializedRanges || !serializedRanges.length) return [];
```
Comment (nschloe) [inline src/js/directives/highlights.js]: `== 0` looks nicer.

File: src/js/directives/highlights.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            range.splitBoundaries();
+
+            // get TextNodes inside the range
+            var textNodes = _.filter(
+              getTextNodes(range.commonAncestorContainer),
+              range.containsNodeText,
+              range
+            );
+
+            // wrap each TextNode in a span to measure it
```
Comment (nschloe) [inline src/js/directives/highlights.js]: Can't we also do it as shown in [1]?

[1] http://stackoverflow.com/a/6966613/353337

File: src/js/directives/highlights.js (hunk 1 above)
Comment (andrenarchy) [inline src/js/directives/highlights.js]: We're concatenating 2 arrays here, so I don't see why we should use `push`.

File: src/js/directives/highlights.js (hunk 2 above)
Comment (andrenarchy) [inline src/js/directives/highlights.js]: We're requiring the controller of the directive `highlightContainer` here. `^^` means: Locate the required controller by searching the element's parents (see the [docs](https://docs.angularjs.org/api/ng/service/$compile)).

File: src/js/directives/highlights.js (hunk 3 above)
Comment (andrenarchy) [inline src/js/directives/highlights.js]: I don't have a strong opinion about this one. It's clear that length always is an integer and testing for zero with `!` is pretty common, isn't it?

File: src/js/directives/highlights.js (hunk 4 above)
Comment (andrenarchy) [inline src/js/directives/highlights.js]: Probably. :) But as noted in your reference, we would have to implement two different versions. I'd prefer a single code for all browsers as a starting point for this functionality. I suggest we add an issue so we can come back to this performance optimization once we have an overview about the performance-critical parts. What do you think?

Comment (nschloe) [inline src/js/directives/highlights.js]: The table at [1] suggests that IE support is in order. Haven't tested though.

[1] http://www.quirksmode.org/dom/w3c_cssom.html

Comment (andrenarchy) [inline src/js/directives/highlights.js]: Tests on [this site](http://www.quirksmode.org/dom/tests/rectangles.html) yield strange results with `getClientRects()` on all major browsers. Example with Chromium:

screenshot from 2015-03-07 16 51 03

**Comment by nschloe** [comment]
Thx, annotations now largely working.

**Pull request merged**